tensorflow_probability/python/experimental/mcmc/covariance_reducer.py

```diff
@@ -0,0 +1,260 @@
+        for population covariance and `ddof=1` for sample covariance. Defaults
+        to the population covariance.
+      name: Python `str` name prefixed to Ops created by this function.
+        Default value: `None` (i.e., 'covariance_reducer').
+    """
+    self._parameters = dict(
+        event_ndims=event_ndims,
+        ddof=ddof,
+        name=name or 'covariance_reducer'
+    )
+
+  def initialize(self, initial_chain_state=None, initial_kernel_results=None):
+    """Initializes a `RunningCovarianceState` using previously defined metadata.
+
+    For calculation purposes, the `initial_chain_state` does not count as a
+    sample. This is a deliberate decision that ensures consistency across
+    sampling procedures (i.e. `tfp.mcmc.sample_chain` follows similar
+    semantics).
+
+    Args:
+      initial_chain_state: A (possibly nested) structure of `Tensor`s or Python
+        `list`s of `Tensor`s representing the current state(s) of the Markov
+        chain(s). It is used to infer the shape and dtype of future samples.
+      initial_kernel_results: A (possibly nested) `tuple`, `namedtuple` or
+        `list` of `Tensor`s representing internal calculations made in a related
+        `TransitionKernel`. For streaming covariance, this argument also has no
+        influence on computation; hence, it is also `None` by default.
+        Nonetheless, this argument is still accepted as it will be supplied by
+        the Streaming MCMC framework.
+
+    Returns:
+      state: `RunningCovarianceState` representing a stream of no inputs and
+        with identical structure to the `initial_chain_state`.
+    """
+    with tf.name_scope(
+        mcmc_util.make_name(self.name, 'covariance_reducer', 'initialize')):
+      initial_chain_state = tf.nest.map_structure(
+          tf.convert_to_tensor,
+          initial_chain_state)
+      self._parameters['shape'] = tf.nest.map_structure(
+          lambda chain_state: chain_state.shape,
+          initial_chain_state)
+      self._parameters['dtype'] = tf.nest.map_structure(
+          lambda chain_state: chain_state.dtype,
+          initial_chain_state)
+      if self.event_ndims is None:
+        self._parameters['event_ndims'] = tf.nest.map_structure(
+            ps.rank, initial_chain_state)
+      elif not nest.is_sequence(self.event_ndims):
+        self._parameters['event_ndims'] = tf.nest.map_structure(
+            lambda _: self.event_ndims,
+            initial_chain_state,
+        )
+      self.strms = nest.map_structure_up_to(
+          # only parameter guaranteed to give desired shallow structure
+          self._parameters['dtype'],
+          sample_stats.RunningCovariance,
+          self.shape, self.event_ndims, self.dtype,
+          check_types=False,
+      )
+      return tf.nest.map_structure(
+          lambda strm: strm.initialize(),
+          self.strms)
+
+
+  def one_step(
+      self,
+      sample,
```

Comment:
chain_state? It's a bit presumptive to call these samples. It's what you called this in initialize as well, so at least we should be consistent across the board.

Reply:
I actually think it's more confusing to call it `chain_state`. I think under the overall framework, it's natural to call these samples (i.e the idea is in `WithReductions` some new sample is taken from the inner kernel then all these `one_step` methods are invoked). Leaving it as `chain_state` might make it less clear that this is supposed to be a new sample? I think in `initialize` it makes sense to refer to it as the initial chain state as it's really not a sample (we don't count it as one). Interested in both what you think about that thought though; might be a good opportunity for learning about better naming strategies. Overall, it's a trivial point that I don't mind changing.

Reply:
Good points! I like `new_chain_state`.